### DISCUSSION TITLE
next in dev environment，component with hooks executed twice

### DISCUSSION THREAD

**Discussion by luxi-record** [vercel/next.js#42563] [Help]
why I used next dev command，then my component with useeffect print some message printed twice？

**Comment by ndejoma** [comment]
To [help you find bugs,](https://beta.reactjs.org/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed) in development React runs setup and cleanup one extra time before the actual setup. This is a stress-test that verifies your Effect’s logic is implemented correctly. If this causes visible issues, your cleanup function is missing some logic. The cleanup function should stop or undo whatever the setup function was doing. The rule of thumb is that the user shouldn’t be able to distinguish between the setup being called once (as in production) and a setup → cleanup → setup sequence (as in development). [See common solutions.](https://beta.reactjs.org/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development)

You can read more here react[ useEffect in development](https://beta.reactjs.org/apis/react/useEffect)

**Reply by icyJoseph** [reply]
Pretty much this....

The big take away is that this is to make sure your components are well behaved. That is to say that you cancel ongoing async work, that you clean up DOM mutations, that you unsubscribe from events, and that you are not doing any funky mutations.

Granted for something like analytics this is a bit of an annoyance, but your dev analytics shouldn't float up to your production analytics. Hopefully projects do have it like that though.
